Folder with null value in userModification/creationDate/modificationD…

blankse opened the pull request:
…ate/userOwner can't be saved

## Changes in this pull request  
Resolves https://github.com/pimcore/pimcore/issues/17019

## Additional info

It is possible to have null in the db columns. Maybe from an old Pimcore versions without typehints or for the user columns from deleted users.

Alternative is to make the db columns not nullable and set Default Values (for Date Fields the current date)

blankse committed:
Folder with null value in userModification/creationDate/modificationDate/userOwner can't be saved

fashxp commented:
hmm, I would prefer not to change the signatures of the methods but to try to fix old values (or investigate the reason, why these values could be null).
WDYT?

blankse commented:
@fashxp So we should create migrations to change the db tables to not nullable?

fashxp commented:
... and figure out some reasonable default values (which could be challenging :smile: ).
At least this is my opinion. What do others think?

blankse commented:
@fashxp I think the two date columns can have the current date as the default value. However, I think the user columns should be nullable. What happens when you delete a user? Do the referenced IDs remain, or will the columns be emptied?

mcop1 commented:
Hello @blankse , just had a quick look and userOwner/userModification get set to `null`. Therefore I´d say the values should be nullable. Since this is already the case in your pr, as far as I can see, it can be merged. Any objections?

blankse commented:
@mcop1 Not from my side :)

fashxp commented:
but creation/modificationDate should not be nullable, right?

mcop1 commented:
@blankse if I got that right, you had the situation that creation/modificationDate was null, right? To make them not nullable, we would need to add a migration for that case.
@fashxp Any other suggestions?

blankse commented:
@mcop1 Yes, right. The alternative is to create a migration to change the db columns. Is it possible to change it to column type TIMESTAMP so we can set the default value to the current timestamp? Otherwise it must be at least `0` as default value instead of `NULL`.

Current
https://github.com/pimcore/pimcore/blob/68892689cc17d299ead2dce18303db8b00da6f76/bundles/InstallBundle/dump/install.sql#L11-L12

Example with Timestamp (db table notifications)
https://github.com/pimcore/pimcore/blob/68892689cc17d299ead2dce18303db8b00da6f76/bundles/InstallBundle/dump/install.sql#L802-L803

Example with 0 value (db table site)
https://github.com/pimcore/pimcore/blob/68892689cc17d299ead2dce18303db8b00da6f76/bundles/InstallBundle/dump/install.sql#L375-L376

blankse commented:
@mcop1 Okay TIMESTAMP column type needs a 'Y-m-d H:i:s' string as value (when I look in the Notification model). So this is not easy to migrate. So we should set the DEFAULT value in the db columns to '0'?

blankse committed:
Merge branch '11.3' into fix_folder_nullable

blankse committed:
Set default value 0 for creationDate and modificationDate columns

blankse commented:
@fashxp @mcop1 I created the migration to set the default value to 0 in the db columns

blankse committed:
Make all setUserOwner/setUserModification nullable

blankse committed:
Make all getUserOwner/getUserModification nullable

mcop1 commented:
Thanks for adding the migration. I will try to run it locally, do some tests and merge it afterwards.

mcop1 commented:
So, pr looks good to me, but on second thought, changing the setters in `ElementInterface` and `AbstractElement` is a breaking change, probably in `CustomLayout` too. 
Guess we need to postpone the pr to Pimcore 12. 
But then we could also think about your second proposal, changing the db types to timestamp and add the current timestamp as default value. Without putting too much thought into it, I would like that solution a bit more.
What do you think?

blankse commented:
@mcop1 Can we merge it for Pimcore 12 now?

blankse committed:
Merge remote-tracking branch 'origin/12.x' into fix_folder_nullable

blankse committed:
Merge remote-tracking branch 'upstream/12.x' into fix_folder_nullable

blankse commented:
@mcop1 Friendly reminder: One month without response :)

mcop1 commented:
Hello @blankse ,
sorry, we talked about the pr internally and I was told you will get a reply, but unfortunately it seems like we forgot or there was a misunderstanding.
I will bring that topic up again, you should get an answer asap.

herbertroth committed:
Remove nullable types for creationDate and modificationDate

Replaced nullable integers with default values of 0 for `$creationDate` and `$modificationDate`. Updated their getters to reflect the non-nullable type, ensuring consistent type usage and preventing potential null-related issues.
